### ISSUE TITLE
AirPcap problem (sometimes)

### ISSUE THREAD

**Issue by gvanem** [gcla/termshark#91]
*  I am running the latest versions of Termshark v2.1.1 and Wireshark.
 *  I checked the [README](https://github.com/gcla/termshark) and [User Guide](https://github.com/gcla/termshark/blob/master/docs/UserGuide.md) and found no answer
 *  I searched [issues](https://github.com/gcla/termshark/issues?q=is%3Aissue) and this has not yet been filed

## Problem 

Launching *Termshark* from directory of Wireshark (x64) like:
```
F:\internet\sniffer\WireShark-3.2\>  start cmd.exe /c termshark_2.1.1_windows_x64\termshark.exe
```

Causes the capture to stop after 1 packet with this error!
![Termshark-Airpcap](https://user-images.githubusercontent.com/945271/89100077-9300b700-d3f4-11ea-9a1a-dfe8bd5c963c.png)

Since a `dumpcap -D` lists:
```
1. \\.\airpcap00 (AirPcap USB wireless capture adapter nr. 00)
2. \Device\NPF_{3A46ACA0-CBED-44BC-A239-6AEA3D0C451D} (Ethernet)
3. \Device\NPF_{7B640089-F33E-4886-823D-87966FA4616F} (Bluetooth-nettverkstilkobling 3)
4. \Device\NPF_{F92984E3-5D40-4AD9-B054-41288EAE699F} (Wi-Fi 2)
```
I assume `\\.\airpcap00` was selected when an adapter wasn't specified.

I'm not sure the above error was a once-in-a-blue-moon event. Termshark works brilliant for the most part 
(although slow for WiFi-traffic like the above). The dark-mode is also way cool!

And the reason I have to use `start cmd.exe /c`, is because my shell (`4NT.exe`) gets killed! during Termshark's run.
I use *ProcessExplorer* and see the parent process vanished while `termshark.exe` was running. 

Highly strange and maybe another issue for later.

I'm on Win-10 (x64).

*EDIT*: I stumbled upon the `C:\Users\Gisle\AppData\Local\termshark\termshark.log` with this:
```
...
time="2020-08-01T12:36:06+02:00" level=info msg="Scheduler requested interface/fifo load for 'AirPcap USB wireless capture adapter nr. 00'"
time="2020-08-01T12:36:06+02:00" level=info msg="Initialized stream widget cache with 100 entries."
time="2020-08-01T12:36:06+02:00" level=info msg="Starting new interface/fifo load 'AirPcap USB wireless capture adapter nr. 00'"
time="2020-08-01T12:36:06+02:00" level=info msg="Starting Iface command: dumpcap.exe [dumpcap -i AirPcap USB wireless capture adapter nr. 00 -w c:\\Users\\Gisle\\AppData\\Local\\termshark\\pcaps\\AirPcap_USB_wireless_capture_adapter_nr._00--2020-08-01--12-36-04.pcap]"
time="2020-08-01T12:36:06+02:00" level=info msg="Read cached tshark fields."
time="2020-08-01T12:36:06+02:00" level=info msg="Started PSML command tshark.exe [tshark -T psml -o gui.column.format:\"No.\",\"%m\",\"Time\",\"%t\",\"Source\",\"%s\",\"Destination\",\"%d\",\"Protocol\",\"%p\",\"Length\",\"%L\",\"Info\",\"%i\" -i - -l --color] with pid 2700"
time="2020-08-01T12:36:06+02:00" level=info msg="Capturing on 'AirPcap USB wireless capture adapter nr. 00'"
time="2020-08-01T12:36:06+02:00" level=info msg="Pcap file c:\\Users\\Gisle\\AppData\\Local\\termshark\\pcaps\\AirPcap_USB_wireless_capture_adapter_nr._00--2020-08-01--12-36-04.pcap has appeared - launching UI"
time="2020-08-01T12:36:06+02:00" level=info msg="Starting Tail command: termshark_2.1.1_windows_x64\\termshark.exe [termshark_2.1.1_windows_x64\\termshark.exe --tail c:\\Users\\Gisle\\AppData\\Local\\termshark\\pcaps\\AirPcap_USB_wireless_capture_adapter_nr._00--2020-08-01--12-36-04.pcap]"
time="2020-08-01T12:36:06+02:00" level=info msg="Launching termshark UI"
time="2020-08-01T12:36:06+02:00" level=info msg="File: c:\\Users\\Gisle\\AppData\\Local\\termshark\\pcaps\\AirPcap_USB_wireless_capture_adapter_nr._00--2020-08-01--12-36-04.pcap"
time="2020-08-01T12:36:06+02:00" level=info msg="Terminal was resized" event="&{{13817447476629760492 3363779901 0x133e6a0} 120 48}"
time="2020-08-01T12:36:07+02:00" level=info msg="Capturing on '-'"
time="2020-08-01T12:36:07+02:00" level=info msg="1 packet captured"

time="2020-08-01T12:36:07+02:00" level=info msg="Did not kill psml process: TerminateProcess: Ingen tilgang."
                                                                                             ^ "No Access"
time="2020-08-01T12:36:08+02:00" level=error msg="Error running command [command: dumpcap.exe [dumpcap -i AirPcap USB wireless capture adapter nr. 00 -w c:\\Users\\Gisle\\AppData\\Local\\termshark\\pcaps\\AirPcap_USB_wireless_capture_adapter_nr._00--2020-08-01--12-36-04.pcap], error: exit status 1]"
```

**Comment by pocc** [comment]
Great work on this detailed writeup!

It looks like dumpcap is not happy with the airpcap interface. Do you have issues capturing on airpcap with Wireshark or plain `dumpcap.exe`? 

You can also change the default capturing utility to tshark in your [termshark config](https://github.com/gcla/termshark/blob/master/docs/UserGuide.md#config-file), which could fix.

**Comment by gvanem** [comment]
> Do you have issues capturing on airpcap with Wireshark or plain dumpcap.exe?

None.

**Comment by pocc** [comment]
1. I just want to verify - so you have tested this with `dumpcap.exe` from the command line and were able to capture >1 packets? This is an additional troubleshooting step.
2. Can you change the `dumpcap` value in your termshark config to tshark like below and see if that fixes?

```
[main]
    dumpcap = "tshark"
```

**Comment by gcla** [comment]
Hi @gvanem - yes, thanks for the report. Strange behavior with 4nt.exe! I've not used that tool before, but I just tried with TCC25 - is that the same program? (I couldn't get TCC26 to install on Windows 10) I watched with process explorer while termshark captured packets, but the shell wasn't killed. Maybe the second issue you reported is fixed in a newer 4nt/TCC?

I can't get usbpcap to cooperate with VirtualBox, so I'll have to steal my wife's laptop later to debug your first issue ;-)

**Comment by gcla** [comment]
Quick update - @pocc - I just tried your suggestion and was able to capture USB packets on Windows :-) Hopefully it clears up your issue, @gvanem

**Comment by gvanem** [comment]
> so you have tested this with dumpcap.exe from the command line and were able to capture >1 packets? 

Yes: When running `f:\internet\sniffer\WireShark-3.2> dumpcap.exe -i1 -c1000 -w air.pcapng` , it runs fine:
```
Capturing on 'AirPcap USB wireless capture adapter nr. 00'
File: air.pcapng
Packets captured: 1000
Packets received/dropped on interface 'AirPcap USB wireless capture adapter nr. 00': 1000/0 (pcap:0/dumpcap:0/flushed:0/ps_ifdrop:0) (100.0%)
```
Also a `start cmd /C termshark_2.1.1_windows_x64\termshark.exe  -r air.pcapng` runs fine.
So it's probably not a problem with decoding 802.11 per se, but maybe how processes and TTYs are used.
(related to how termshark  does'nt like my 4NT shell. Hence the use of `start cmd /C` ..)

> [main]
>     dumpcap = "tshark"

Same as before mostly:
![Termshark-Airpcap-2](https://user-images.githubusercontent.com/945271/89117587-38706500-d49f-11ea-86e7-ac0bf4d61c11.png)

And when that occured, `termshark.log` has:
```
...
time="2020-08-02T08:47:44+02:00" level=info msg="Launching termshark UI"
time="2020-08-02T08:47:44+02:00" level=info msg="Terminal was resized" event="&{{13817525534968680144 5771012301 0x133e6a0} 120 48}"
time="2020-08-02T08:47:51+02:00" level=info msg="0 packets captured"
time="2020-08-02T08:47:51+02:00" level=info msg="Did not kill psml process: TerminateProcess: Ingen tilgang."
time="2020-08-02T08:47:51+02:00" level=error msg="Error running command [command: tshark.exe [tshark -i Ethernet -w c:\\Users\\Gisle\\AppData\\Local\\termshark\\pcaps\\Ethernet--2020-08-02--08-47-40.pcap], error: exit status 1]"
time="2020-08-02T08:47:51+02:00" level=info msg="Did not kill tail process: invalid argument"
```

But a `start cmd /C termshark_2.1.1_windows_x64\termshark.exe  -i 1`  (my `\\.\airpcap00`), works sometimes?!

I did some experimenting with  `AirpcapConf`  (the airpcap control program) and changing Capture-type among:
  * 802.11 Only
  * 802.11 + Radio and  
  * 802.11 + PPI.

But that didn't seems to have no effect.

**Comment by gvanem** [comment]
>  I've not used that tool before, but I just tried with TCC25 - is that the same program? 

4NT is the *grand-father* of [TakeCommand](https://jpsoft.com/products/take-command.html) and [TCC](https://jpsoft.com/products/tcc-cmd-prompt.html). Has worked for me for the most part since I installed Win-XP in 2004.

**Comment by gcla** [comment]
Hi @gvanem - thanks for the update. I'm wondering how to debug further... (I don't have an airpcap adapter, unfortunately)

Have you ever used Process Monitor on Windows? 

https://docs.microsoft.com/en-us/sysinternals/downloads/procmon

It's a single binary, so you can just drop it on your desktop and run it. The way I used it yesterday was to 

- set a filter that looked like:
```
Operation  Is  Process Create   Include
Operation  Is  Process Exit     Include
```
- start capturing events (ctrl-e)
- then run termshark in TCC/powershell/cmd

Those filtered events will show you which processes are kicked off along with their flags and eventual exit codes. You can save the log as a PML file which might be useful to see. If your machine is quiet, you might only get events from termshark, tshark and maybe dumpcap (or you could could filter out other processes)

I understand if you don't want to try this - it's giving you work to do.

**Comment by gvanem** [comment]
I had much more trouble now getting the failure as was more frequent earlier. 

But I used `AirpcapConf.exe` to include **All Frames** (*the Frame Check Sequence* filter?). 
I fiddled with changing the channels and radio-capture types, but just now `dumpcap` and `termcap` 
behaves w/o any error like previously.

But for reference, the PML file from ProcMon on  [Success](http://www.watt-32.net/misc/procmon-airpcap-success.pml).
